File: benchmarks-ocaml/owl_explosion.ml
open Interpreter

let owl_combinator = Lambda (Lambda (Apply (tvar 0, Apply (tvar 1, tvar 0))))

let rec owl_explosion depth =
  assert (depth > 0);
  let rec go i acc =
    if i >= depth then acc else go (i + 1) (Apply (acc, owl_combinator))
  in
  go 1 owl_combinator

let benchmark_term = Apply (owl_explosion 10000, owl_combinator)

let () = ignore (Sys.opaque_identity (denote [] benchmark_term))

File: benchmarks-haskell/owl-explosion.hs
import Control.Exception (assert)
import Interpreter

owlCombinator :: Term
owlCombinator = Lambda (Lambda (Apply (TVar 0) (Apply (TVar 1) (TVar 0))))

owlExplosion :: Int -> Term
owlExplosion depth =
  assert (depth > 0) $ go 1 owlCombinator
  where
    go i acc
      | i >= depth = acc
      | otherwise = go (i + 1) (Apply acc owlCombinator)

benchmarkTerm :: Term
benchmarkTerm = Apply (owlExplosion 1000000) owlCombinator

main :: IO ()
main = print $ case denote [] benchmarkTerm of
  VClosure {} -> "VClosure"
  VUnaryOp {} -> "VUnaryOp"
  VBinaryOp {} -> "VBinaryOp"
  VBinaryOpAux {} -> "VBinaryOpAux"
  VConst n -> show n
